src/V3EmitMk.cpp

```diff
@@ -31,11 +31,557 @@ class EmitMk final {
+
+        bool isConcatenatingFile() const { return !concatenatedFileNames.empty(); }
+    };
+
+    struct FileNameWithScore final {
+        std::string name;
+        std::int64_t score;
+    };
+
+    // Data of a single work unit used in `singleConcatenatedFilesList()`.
+    struct WorkList final {
+        std::int64_t totalScore{0};
+        std::vector<FileNameWithScore> files{};
+        // Number of buckets assigned for this list. Used only in concatenable lists.
+        int bucketsNum{0};
+        // Indicated whether files on this list can be concatenated.
+        bool isConcatenable{true};
+        // Work list ID for debugging.
+        int dbgId;
+
+        WorkList() = delete;
+        WorkList(int id)
+            : dbgId{id} {}
+    };
+
+    // Debug logging helper: Returns properly indented "* " string.
+    // Used as a prefix of log lines representing list entry.
+    static std::string logMsgPrefixListEntry(unsigned level, unsigned ancestorsWithStatusNum = 0) {
+        UASSERT(level >= ancestorsWithStatusNum,
+                "Level must include number of ancestors with status");
+        const int spacesNum = ancestorsWithStatusNum * 6 + (level - ancestorsWithStatusNum) * 2;
+        const std::string indentSpaces = std::string(spacesNum, ' ');
+        return indentSpaces + "* ";
+    };
+
+    // Debug logging helper: Returns properly indented "* [+] " if status is true,
+    // otherwise "- [ ]". Used as a prefix of log lines representing list entry.
+    static std::string logMsgPrefixListEntryWithStatus(bool status, int level,
+                                                       int ancestorsWithStatusNum = 0) {
+        return logMsgPrefixListEntry(level, ancestorsWithStatusNum) + (status ? "[+] " : "[ ] ");
+    };
+
+    // Debug logging: prints scores histogram
+    static void debugLogScoreHistogram(const std::vector<std::int64_t>& sortedScores) {
+        constexpr int LOG_LEVEL = 6;
+
+        if (debug() < LOG_LEVEL) return;
+
+        constexpr int MAX_BAR_LENGTH = 80;
+        constexpr int MAX_INTERVALS_NUM = 60;
+
+        const auto topScore = sortedScores.back();
+        UINFO(LOG_LEVEL, "Top score: " << topScore << endl);
+        const int maxScoreWidth = topScore < 10       ? 1
+                                  : topScore < 100    ? 2
+                                  : topScore < 1000   ? 3
+                                  : topScore < 10000  ? 4
+                                  : topScore < 100000 ? 5
+                                                      : 7;
+
+        const int intervalsNum = std::min<int>(topScore + 1, MAX_INTERVALS_NUM);
+        const double intervalWidth = double(topScore + 1) / intervalsNum;
+
+        struct Interval final {
+            int64_t lowerBound = 0;
+            int size = 0;
+        };
+
+        std::vector<Interval> intervals;
+        intervals.resize(intervalsNum);
+
+        for (int64_t score = topScore; score >= 0; --score) {
```

Comment:
Shouldn't it be a uint64_t?

Reply:
I find using signed integer for this (and many other values) better, as it prevents some types of bugs from being introduced. Conveniently, the referenced loop is a good example - it would never end with unsigned integer, since the stop condition `score >= 0` would always be true.
In many cases if there is a risk that 63 bits is not enough to store a value, using 64 bits is likely not enough either.

Reply:
I changed the condition of the `for` loop and then I changed the type to uint64_t.